web/web_test.go

```diff
@@ -0,0 +1,1868 @@
+				ColDefs: map[string]ddl.ColumnDef{
+					"a": ddl.ColumnDef{Name: "a", T: ddl.Type{Name: ddl.String, Len: ddl.MaxLength}},
+					"b": ddl.ColumnDef{Name: "b", T: ddl.Type{Name: ddl.String, Len: ddl.MaxLength}},
+					"c": ddl.ColumnDef{Name: "c", T: ddl.Type{Name: ddl.Int64}},
+					"d": ddl.ColumnDef{Name: "d", T: ddl.Type{Name: ddl.String, Len: int64(6)}},
+					"e": ddl.ColumnDef{Name: "e", T: ddl.Type{Name: ddl.Numeric}},
+					"f": ddl.ColumnDef{Name: "f", T: ddl.Type{Name: ddl.Timestamp}},
+					"g": ddl.ColumnDef{Name: "g", T: ddl.Type{Name: ddl.Int64}},
+					"h": ddl.ColumnDef{Name: "h", T: ddl.Type{Name: ddl.String, Len: int64(1)}},
+					"i": ddl.ColumnDef{Name: "i", T: ddl.Type{Name: ddl.Bytes, Len: ddl.MaxLength}},
+					"j": ddl.ColumnDef{Name: "j", T: ddl.Type{Name: ddl.Date}},
+					"k": ddl.ColumnDef{Name: "k", T: ddl.Type{Name: ddl.Float64}},
+					"l": ddl.ColumnDef{Name: "l", T: ddl.Type{Name: ddl.Int64}},
+					"m": ddl.ColumnDef{Name: "m", T: ddl.Type{Name: ddl.Int64}},
+					"n": ddl.ColumnDef{Name: "n", T: ddl.Type{Name: ddl.String, Len: ddl.MaxLength}},
+					"o": ddl.ColumnDef{Name: "o", T: ddl.Type{Name: ddl.Timestamp}},
+					"p": ddl.ColumnDef{Name: "p", T: ddl.Type{Name: ddl.Int64}},
+				},
+				Pks: []ddl.IndexKey{ddl.IndexKey{Col: "a"}},
+			},
+			expectedIssues: map[string][]internal.SchemaIssue{
+				"a": []internal.SchemaIssue{internal.Widened},
+				"b": []internal.SchemaIssue{internal.Widened},
+				"c": []internal.SchemaIssue{internal.Widened},
+				"g": []internal.SchemaIssue{internal.Serial},
+				"l": []internal.SchemaIssue{internal.Widened},
+				"m": []internal.SchemaIssue{internal.Serial},
+				"o": []internal.SchemaIssue{internal.Timestamp},
+				"p": []internal.SchemaIssue{internal.Widened},
+			},
+		},
+		{
+			name: "Test bad request",
+			payload: `
+    {
+      	"bool":"INT64",
+		"int8":"STRING",
+		"float4":"STRING",
+    }`,
+			statusCode: http.StatusBadRequest,
+		},
+	}
+	for _, tc := range tc {
+		app.driver = "postgres"
+		app.conv = internal.MakeConv()
+		buildConvPostgres(app.conv)
+		payload := tc.payload
+		req, err := http.NewRequest("POST", "/typemap/global", strings.NewReader(payload))
+		if err != nil {
+			t.Fatal(err)
+		}
+		req.Header.Set("Content-Type", "application/json")
+		rr := httptest.NewRecorder()
+		handler := http.HandlerFunc(setTypeMapGlobal)
+		handler.ServeHTTP(rr, req)
+		var res *internal.Conv
+		json.Unmarshal(rr.Body.Bytes(), &res)
+		if status := rr.Code; int64(status) != tc.statusCode {
+			t.Errorf("handler returned wrong status code: got %v want %v",
+				status, tc.statusCode)
+		}
+
+		if tc.statusCode == http.StatusOK {
+			assert.Equal(t, tc.expectedSchema, res.SpSchema["t1"])
+			assert.Equal(t, tc.expectedIssues, res.Issues["t1"])
+		}
+	}
+
+}
+
+func TestGetConversionPostgres(t *testing.T) {
+	app.driver = "postgres"
+	app.conv = internal.MakeConv()
+	buildConvPostgresMultiTable(app.conv)
```

Comment:
I think we are testing too many things in this unit test, I imagine it would be quite hard to maintain this test case (and similarly for Mysql). 

- we are testing handler functionality
- we are testing whether 'rateSchema' function
- we are testing functionality from report.go

We should instead test these as follows:

- Instead of adding all this code in buildConvPostgresMultiTable, I think we should just reuse buildConvPostgres here and check that the status code returned is 200 OK and that expectedConversion contains keys for all the tables.

- All the work to add buildConvPostgresMultiTable is not lost though and this function could move to a new test in report_test.go and that should just test that the report generated for these set of tables is as expected

- Lastly, I don't see a lot of value in testing 'rateSchema' function since the test would just duplicate what is there in the function.

At any rate, to make the test maintainable, we should get rid of buildConvPostgresMultiTable and matching colors in this test.

Reply:
Same comment for corresponding mysql test and utility function.

Reply:
The reason why i have created buildConvPostgresMultiTable, is to test whether getConversionRate returns correct colors for rate of conversion. Using buildConvPostgres would not be good for testing rate of conversion, as it is highly used for testing types only.
Also i didn't understand the first point, are you saying we should just test wether expectedConversion contains the keys for the tables only? not the color codes?

Reply:
The issue here is that if I change the code that creates the report and/or the rating logic, then the color codes can change. But looking at the input data for the test i.e., buildConvPostgresMultiTable, it will be quite tricky to figure out why. The reason being that there are multiple conversions going on here - 

1. buildConvPostgresMultiTable to report
2. report to rating
3. rating to color

We should have unit test for each of the conversion piece since that's easier to reason about and maintain. What we have right now is basically an end to end integration test for rating a given input schema. For example, looking at the current test case, I have no way to tell whether the test is correct or not.

> Also i didn't understand the first point, are you saying we should just test wether expectedConversion contains the keys for the tables only? not the color codes?

Yes, that will verify that the API layer is working as intended since that should be the objective for this unit test. If you want you can have a test case for rateSchema to check for colors for a given rating/warnings etc, but that would be same as the source code itself.

Reply:
Understood, i'll update the test case for TestGetConversionPostgres and similarly for MySQL, to test the keys it contains, and for that we probably don't require buildConvPostgresMultiTable, so i'll remove it. but i don't think we should separately test rateSchema as it is just switch case.

Reply:
Implemented this change, removed buildConvPostgresMultiTable and buildConvMySQLMultiTable as they are not required, and test for TestGetConversionPostgres and TestGetConversionMySQL is updated to test the keys the result contain.